### ISSUE TITLE
Native support for sending responses in arrays instead of objects (Query results as array of arrays)

### ISSUE THREAD

**Issue by ozum** [prisma/prisma#17576]
### Discussed in https://github.com/prisma/prisma/discussions/17554

<sup>Originally posted by **ozum** January 26, 2023</sup>

Hi,

## TLDR;

Having data as an array of arrays is much more efficient for some use cases.

## What?

Prisma, like most ORM's returns query results as array of objects:

`prisma.product.findMany();`

```js
[
  { id: 1, color: "red", ... , isActive: true },
  { id: 2, color: "green", ..., isActive: true },
]
```

I want to have results as array of arrays:

```js
[
  [1, "red", ...., true],
  [2, "green", ...., true],
]
```

For example, node-postgres has [row-mode](https://node-postgres.com/features/queries#row-mode), Perl DBI has [selectall_arrayref()](https://metacpan.org/pod/DBD::Pg#selectall_arrayref) method for this.

## Rationale:

For large data sets an array of objects "[{}, {}]" has some disadvantages, because of unnecessary column names repeating in each object.

* Higher data size transferred between server and client.
* Higher memory consumption in the client.
* Higher CPU time to process large data set. (Maybe!!! This depends of further processing the data nad usage model)

## Existing Optimizations:

I'm very well aware of existing optimizations:

✅ Compression (gzip, brotli, etc.) works very well for repeating texts which reduce data transferred significantly.
  * Takes extra CPU time.
  * Still worse than an array of arrays. The best compression is the lack of data.

✅ V8 memory optimizations.
  * V8 cannot be sure all current data (db rows) and future data (db rows) are in the same structure.

## Workarounds

* Convert data to an array of arrays on the server manually or using a middleware.

DBMS already presents data as an array of arrays (or equivalent). DB drivers usually process the data to convert it to an array of objects. Converting it again to an array of arrays is redundant work. Even maybe, sometimes it will be converted to an array of objects again.

## Further Exploration

How to handle & use data on the client side efficiently and as much as developer-friendly supporting TypeScript? This can be a thought exercise...

* Classes

```ts
class Product {
  private rowData: { id: number, name: string, .... };
  
  get id() {
    return this.rowData[0]
  }

  get name() {
    return this.rowData[2]
  }
}
```

* A class extending Array

```ts
class KeyedArray<T> extends Array {
  constructor(keyList: string[], data: T[] ) {
  }
}

class Product<T> extends KeyedArray<T> {

}
```

* Accessing arrays using const variables:

```js
const id = 0;
const color = 1;
const name = 2;

data.forEach((row) => {
  row[color]
});
```

**Comment by janpio** [comment]
How would that work with Prisma's type system?

**Comment by kavyantic** [comment]
This is an extremely necessary feature as it would also allow the drizzle orm to have full [plug-in](https://orm.drizzle.team/docs/prisma#limitations) support for querying database which allows very smooth migration from prisma to drizzle for sake of performance.

**Comment by kavyantic** [comment]
> How would that work with Prisma's type system?

Type system can be kept away for this case.

**Comment by Ustice** [comment]
As long as your either provide or know the property names, this becomes totally possible.

[Here's an example](https://www.typescriptlang.org/play/?#code/PTAEEEGdQGwewHYHNQENoE84FdQGsE4B3UAFwAsBTUAYzhmwFsF9KNIAaUAKjKpYCWpUAOgAHOJEgCARjGqk4oACZKKo2nEZj5pSjAwiWqMhjGUAtJFQAzakVQYAdACgXpM9QAKk4QF5QAG8XUDIheQAuUEhSACcBZBDQGThlDCiY+MTQgWUohCYZSliksWw5USo80AARVD0k7DFlespquoaAXzc6BBjQCRjIKJ8YgG0AXVAAsaTg0NDScMoogCIAMQFY-sHSVY4k0JS0tc3t0gBCfcORaoAGA4WB8phKtvzKEg7KAAoASkeCyaLT01QQn1qrX+gM6gPmCyWpEioFWAGVKL1lANfNcnsd0ij0ZiroCctUAIyk54VSBVD5fKEAm7A1pgiHfaFJboTHqIfqUbQeaZBTpoaC9GK8vrCXYAYXoTAQAHlYspisKxqAAOSI+Rarha3L67VlGlVY1almgi34rWgKboTTS6L1UQ2ASUaA-PBsOA2UCjUh-SZuGzYBA0JaIUCEUgAJXDCASKH+QSSEuERFiiCQgflDGYKrVsQ12t1lAtRoNptetLa4FIFqtbTtDvFfOE1iWkHdntA3t9-sDwZ5oRATwnk6n05nE4AeqBZahsEhyMIhC5uu5PAGBDQfcoACpNeSgAA8h64x501EoAA89AhlF6fVh-YfQAAyaJxZMj0BcKipCtKA96Ps+aAIBgkwahMAB8wpJDBYGUE+0DXqeAD8oBAa0SRRBht4PqhEGagkdglmc-LEWhrBvqAH7fpkyZcE4bHkeqcaesIKG0QO9GMT+WRIP+o4LNhXh7gehEXlwXExFwmpsU4uF6FeYxUaQEzwTcHwAG7FG4HjmAGviEcKkn7m0MmBlwxmUH62IxPmipFsUcEuOOoRzphnlgAAko20BEHAsR4MmaDCBQCieBY8gGTAXAyNgmbUOCbRkGosSoH0NihYwoA4CWIImIocDKaG4aRgI0akCeNiGLJi4KswAByqCMH2vEQfxjmCcxyAjnBPyoLE2UYEqNhKjIABWGKkMMDGTFwdAFgg7WdYtLltR1np-GmoSxJQdWxMYY2OJN01zZGkBOIwqBiD8Nw-EIAr7X4CGrYqG2endD1PZOPxfTtnXvQhr2MGMwPrbtYmhPtjqWdJJ6UE120w5tHnw5uUo7GZKPQAEdU6A1Py7JwTmkOjbmxH8bjjoFWrQKIWihWI5B7tESjwMgYqgFguAJDQDBqqBBmxIYr5GTuABaAhiIRkBQAAoggv6emeADSbAUwAaqgDCelwqvq4ToCTAhARJNr7CgTREEDUgIZPNh+uG9A3XQDl0FTDcoTYSb8Rdfb0BjI7XDexMztTthNuQGMWryMgFBalMnugG72CegnSdICncMx+bach6Acd+4XgcepA5eTlEcd2+BodGBRoCaTbrHsQgLfyaQk31wXs6uwbWceyXZFd+qmmZ5QHdOBxJY95N0-QAPs6gNhcsKwTKtq0HkBnovNhx3J3FL8PRvm6AymVxfmpt2wXBT+f9r2lja9PPphnv6En8lLOURakVqARg2B+jkFQAZPg1BrCdVgKhPO5AtQ11AAAm+0AQH9CKAMVAWxoCOVfF7J8oA9Ln0gEg2u2pl7ANAcILB3tJjkIWAA+uGDaHQN-MgBhbhtwmTQYeOAV15oXhRpNNBV5hT2UcgKMQHhLagCQsXRuDERE2DQTcbCh5dLKJvKI3eVcG4kVDuPFuNsDG0XDs3dU097Sz3nqAQ+XgcHbHtOXAOejPT8MEZGA+p8bCONwVwQI8jZxjE1kYOifVQAAB9S5sAmHXMxEFTHIJdhnZ+KSmGl0SdAV8kSMmpMPKE3238UExkoOLYJU5Ohvw-mU8W0sTKK08bNIRx80nuzkXwgRLTvGb0VjvU2WsdZcGXnBDyPDqCHioEdcAT4ABCqB9zgCQDglgAQmndOuqQM8kj-RyhasqVUxQuCBkIuM+yoAACy9QaBUDNlM4olBZnKAWUslZCRsmmX6NhOIWdSk2ANpASgfkFg+UqhGKMLBibyEgM0rZ550wdkhMBT54YCDEAQJMZalSMzNTWj9UeSjHbLSSFxE6fRPkbK8ds9GBKuAdFQGMAojAiixB0ghH4JUogMq4AgXaW0DkEv2vCI65KVD1FQH9R6nL6hMDBjGflTgjrKGwDQX4z04CzV5btLgCQ1R3nlT8IJylNUzUUnyzq8TxV1Uhnq+8UxOgI2gGS7Ap0KY3GNWxaRQpRSOhdW6pIfw6ZblxdFGZ8zFl4GWasiRBM4XzTJvjG8FN9lrRpnTLy3lfK43oJQJw8AkA-FWOZcW0hED7FAAAKVREqVqThHYCFJrsRWvLsAwESqAAATEGlwEpc35rgIW1Y1LiHFDLQgCt1ba31o4UgRtGAfhhqeRGt5qzW3tq4N2v4QA). It's not optimized, but more for illustration, and playing around with the idea. This isn't strictly working with Prisma's type system, but more generic and able to be modified to do so.
